### DISCUSSION TITLE
AttributeError: 'Detect' object has no attribute 'nm'

### DISCUSSION THREAD

**Discussion by Sheen23** [ultralytics/yolov5#10182] [Q&A]
when I want to train segment/train.py, this problem comes out. how can I solve it...please
![image](https://user-images.githubusercontent.com/95073029/202373246-9d74afe1-c495-4285-959f-4d0fadb25b75.png)

**Comment by glenn-jocher** [comment]
@Sheen23 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

Also see the Segmentation notebook for examples:
https://github.com/ultralytics/yolov5/blob/master/segment/tutorial.ipynb

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by Wouter1** [comment]
Same problem here. 

This how-to-reproduce is pretty complex unfortunately as you need so many files and handwork is involved in configuring. 



I have 5 train images.
I created a dataset using labelme.

Then I used labelme2yolo to create a yolo dataset.
cd data_annotated
labelme2yolo --json_dir .


I copied the dataset to yolov5/datasets/YOLODataset


dataset.yaml inside there:

```
train: ./YOLODataset/images/train/
val: ./YOLODataset/images/val/
test: ./YOLODataset/images/test/
nc: 9
names: ['9 bracket', '7 plate', '6 bar', '3 sphere', '8 transparentplate', '1 pyramid', '4 halfsphere', '2 cube', '5 ring']
```



I copied dataset.yaml to the root of the yolov5 folder. Actually it's not clear where you should run yolo5 and what pythonpath should be set at. So I run from root of yolov5 and make all paths relative to there

So I make a root copy of dataset.yaml with modified content
```
train: datasets/YOLODataset/images/train/
val: datasets/YOLODataset/images/val/
test: datasets/YOLODataset/images/test/
nc: 9
names: ['9 bracket', '7 plate', '6 bar', '3 sphere', '8 transparentplate', '1 pyramid', '4 halfsphere', '2 cube', '5 ring']
```


Then I creates a models/set1.yaml file with this
```
# YOLOv5 🚀 by Ultralytics, GPL-3.0 license

# Parameters
nc: 9  # number of classes
depth_multiple: 0.33  # model depth multiple
width_multiple: 0.50  # layer channel multiple
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32

# YOLOv5 v6.0 backbone
backbone:
  # [from, number, module, args]
  [[-1, 1, Conv, [64, 6, 2, 2]],  # 0-P1/2
   [-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
   [-1, 3, C3, [128]],
   [-1, 1, Conv, [256, 3, 2]],  # 3-P3/8
   [-1, 6, C3, [256]],
   [-1, 1, Conv, [512, 3, 2]],  # 5-P4/16
   [-1, 9, C3, [512]],
   [-1, 1, Conv, [1024, 3, 2]],  # 7-P5/32
   [-1, 3, C3, [1024]],
   [-1, 1, SPPF, [1024, 5]],  # 9
  ]

# YOLOv5 v6.0 head
head:
  [[-1, 1, Conv, [512, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 6], 1, Concat, [1]],  # cat backbone P4
   [-1, 3, C3, [512, False]],  # 13

   [-1, 1, Conv, [256, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 4], 1, Concat, [1]],  # cat backbone P3
   [-1, 3, C3, [256, False]],  # 17 (P3/8-small)

   [-1, 1, Conv, [256, 3, 2]],
   [[-1, 14], 1, Concat, [1]],  # cat head P4
   [-1, 3, C3, [512, False]],  # 20 (P4/16-medium)

   [-1, 1, Conv, [512, 3, 2]],
   [[-1, 10], 1, Concat, [1]],  # cat head P5
   [-1, 3, C3, [1024, False]],  # 23 (P5/32-large)

   [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
  ]
```

It's basically a copy of  a yolov.yaml file with the nc modified.


Now you can go to root of yolov5 and run

python segment/train.py --img 640 --batch 16 --epochs 5 --data dataset.yaml --cfg models/set1.yaml 

Some example from your wiki also used a --weights option. Some other example specifies --weights but without argument, and that one fails right away. I found that completely leaving out --weights at least starts the training.


That will give you the reported error
```
$ python segment/train.py --img 640 --batch 16 --epochs 5 --data dataset.yaml --cfg models/set1.yaml --weights ''
segment/train: weights=, cfg=models/set1.yaml, data=dataset.yaml, hyp=data/hyps/hyp.scratch-low.yaml, epochs=5, batch_size=16, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, bucket=, cache=None, image_weights=False, device=, multi_scale=False, single_cls=False, optimizer=SGD, sync_bn=False, workers=8, project=runs/train-seg, name=exp, exist_ok=False, quad=False, cos_lr=False, label_smoothing=0.0, patience=100, freeze=[0], save_period=-1, seed=0, local_rank=-1, mask_ratio=4, no_overlap=False
github: up to date with https://github.com/ultralytics/yolov5 ✅
YOLOv5 🚀 v7.0-53-g65071da Python-3.10.6 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 2070, 7967MiB)

hyperparameters: lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0
TensorBoard: Start with 'tensorboard --logdir runs/train-seg', view at http://localhost:6006/

                 from  n    params  module                                  arguments                     
  0                -1  1      3520  models.common.Conv                      [3, 32, 6, 2, 2]              
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]                
  2                -1  1     18816  models.common.C3                        [64, 64, 1]                   
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]               
  4                -1  2    115712  models.common.C3                        [128, 128, 2]                 
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]              
  6                -1  3    625152  models.common.C3                        [256, 256, 3]                 
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]              
  8                -1  1   1182720  models.common.C3                        [512, 512, 1]                 
  9                -1  1    656896  models.common.SPPF                      [512, 512, 5]                 
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]              
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 12           [-1, 6]  1         0  models.common.Concat                    [1]                           
 13                -1  1    361984  models.common.C3                        [512, 256, 1, False]          
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]              
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 16           [-1, 4]  1         0  models.common.Concat                    [1]                           
 17                -1  1     90880  models.common.C3                        [256, 128, 1, False]          
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 19          [-1, 14]  1         0  models.common.Concat                    [1]                           
 20                -1  1    296448  models.common.C3                        [256, 256, 1, False]          
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]              
 22          [-1, 10]  1         0  models.common.Concat                    [1]                           
 23                -1  1   1182720  models.common.C3                        [512, 512, 1, False]          
 24      [17, 20, 23]  1     37758  models.yolo.Detect                      [9, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
set1 summary: 214 layers, 7043902 parameters, 7043902 gradients, 16.0 GFLOPs

AMP: checks passed ✅
optimizer: SGD(lr=0.01) with parameter groups 57 weight(decay=0.0), 60 weight(decay=0.0005), 60 bias
train: Scanning /home/wouter/Desktop/yolov5/datasets/YOLODataset/labels/train.cache... 5 images, 0 backgrounds, 
val: Scanning /home/wouter/Desktop/yolov5/datasets/YOLODataset/labels/val.cache... 2 images, 0 backgrounds, 0 co

AutoAnchor: 5.23 anchors/target, 1.000 Best Possible Recall (BPR). Current anchors are a good fit to dataset ✅
Plotting labels to runs/train-seg/exp15/labels.jpg... 
Traceback (most recent call last):
  File "/home/wouter/Desktop/yolov5/segment/train.py", line 658, in <module>
    main(opt)
  File "/home/wouter/Desktop/yolov5/segment/train.py", line 554, in main
    train(opt.hyp, opt, device, callbacks)
  File "/home/wouter/Desktop/yolov5/segment/train.py", line 254, in train
    compute_loss = ComputeLoss(model, overlap=overlap)  # init loss class
  File "/home/wouter/Desktop/yolov5/utils/segment/loss.py", line 40, in __init__
    self.nm = m.nm  # number of masks
  File "/home/wouter/Desktop/yolov5/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1269, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'Detect' object has no attribute 'nm'. Did you mean: 'm'?
```


I hope I got everything,I tried a lot of things

**Reply by Polaris-Zhang** [reply]
hey bro,now,I have face the same question:
AutoAnchor: 2.77 anchors/target, 1.000 Best Possible Recall (BPR). Current anchors are a good fit to dataset 
Plotting labels to ..\runs\train-seg\exp25\labels.jpg... 
Traceback (most recent call last):
  File "D:\DL\yolov5-7.0\yolov5-7.0\segment\train.py", line 658, in <module>
    main(opt)
  File "D:\DL\yolov5-7.0\yolov5-7.0\segment\train.py", line 554, in main
    train(opt.hyp, opt, device, callbacks)
  File "D:\DL\yolov5-7.0\yolov5-7.0\segment\train.py", line 254, in train
    compute_loss = ComputeLoss(model, overlap=overlap)  # init loss class
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\DL\yolov5-7.0\yolov5-7.0\utils\segment\loss.py", line 40, in __init__
    self.nm = m.nm  # number of masks
              ^^^^
  File "C:\Users\15401\.conda\envs\pytorch\Lib\site-packages\torch\nn\modules\module.py", line 1614, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'Detect' object has no attribute 'nm'. Did you mean: 'm'?
thus，Have you solved this problem? If so, could you please reply to me? thanks!

**Comment by Zerohertz** [comment]
I also got the same error.

`--cfg` makes that problem!

Exclude `--cfg` when running segment/train.py

**Comment by jehan88** [comment]
i had the same problem then i solved it by using --weights yolov5s-seg.pt

**Reply by acderem123** [reply]
thank you very much

**Comment by WtElio** [comment]
The same error happened to me,The following is my solution:
make sure your model.yaml is a segment yaml file
![image](https://github.com/ultralytics/yolov5/assets/133865805/3e752999-074c-4e86-bea7-c391f4286a0a)

**Reply by Tina291** [reply]
Thanks,it's helpful
